### PR TITLE
Add weights to AMCL particle cloud

### PR DESCRIPTION
**Pull request by naiveHobo**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1623  |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | TurtleBot3 gazebo simulation |

---

## Description of contribution in a few bullet points

* Adds weight information to AMCL particle cloud 
* Add Particle and ParticleCloud msgs

---

## Future work that may be required in bullet points

* Find a better place for the Particle and ParticleCloud msgs instead of nav2_msgs
* Remove the PoseArray publisher in AMCL node if visualization is not needed
* Add an rviz plugin to visualize ParticleCloud msg which scales pose according to weight

### CODE REVIEW

**Comment by naiveHobo** [comment]
Currently, the particle cloud is published as a PoseArray which enables visualization of the particle cloud in RViz (which also happens to be the only subscriber for the particle cloud currently).

Publishing the particle poses and weights together would require a custom message. As far as I can tell, there's no standard message with such a definition. Hence, the addition of the Particle and ParticleCloud msgs.

One problem with this is that the ParticleCloud can't be visualized in RViz. This means that the AMCL node would have to publish the PoseArray message like it does right now if visualization is necessary, along with another topic to publish the particle cloud with scores using this custom message.

Another way would be to publish the weights and poses as separate messages on different topics. The poses can still be visualized in RViz as PoseArray msgs and scores would have to be published as a Float64 array. However, if the purpose of adding weights is debugging and quality analytics, I'd argue that the score information needs a timestamp. Again, such a message is not available in the standard message definitions and would require a custom message.

I have added the Particle and ParticleCloud messages to nav2_msgs right now. I'm not sure if that's the best place to add these messages since they are not universal to the navigation2 stack.

**Comment by shrijitsingh99** [comment]
For visualization I think best would be [Markers](http://wiki.ros.org/rviz/DisplayTypes/Marker) of Arrow type, you can scale the arrow according to the weight.

**Comment by SteveMacenski** [comment]
@shrijitsingh99 the best solution is probably supporting the particle cloud type in rviz itself. Because its a pretty keystone package in ROS so it should have first-class support. It shouldn't be too hard to implement, it mostly involves just starting with the PoseArray plugin and modifying it to take in the new topic and potentially conveying the weight information with color, size, visibility, etc.

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation2/issues/1678 new ticket for the plugin. I think that's good for this PR. I'm OK to merge. @shivaang12 thoughts?

File: nav2_msgs/msg/Particle.msg [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# This represents an individual particle with weight produced by a particle filter
+
+geometry_msgs/Pose pose
+
+float64 weight
```
Comment (SteveMacenski) [inline nav2_msgs/msg/Particle.msg:5]: Ah small, add new lines at end of files

File: nav2_msgs/msg/ParticleCloud.msg [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# This represents a particle cloud containing particle poses and weights
+
+std_msgs/Header header
+
+# Array of particles in the cloud
+Particle[] particles
```
Comment (SteveMacenski) [inline nav2_msgs/msg/ParticleCloud.msg:6]: Ah small, add new lines at end of files

File: nav2_amcl/src/amcl_node.cpp [hunk 3]
```diff
@@ -266,6 +266,7 @@ AmclNode::on_activate(const rclcpp_lifecycle::State & /*state*/)
   // Lifecycle publishers must be explicitly activated
   pose_pub_->on_activate();
   particlecloud_pub_->on_activate();
+  particlecloudwithweights_pub_->on_activate();
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:269]: In the `on_configure` method can you do a RCLCPP_WARNING on the deprecation of the PoseArray topic?

File: nav2_amcl/src/amcl_node.cpp [hunk 4]
```diff
@@ -1242,6 +1255,10 @@ AmclNode::initPubSub()
     "particlecloud",
     rclcpp::SensorDataQoS());
 
+  particlecloudwithweights_pub_ = create_publisher<nav2_msgs::msg::ParticleCloud>(
+    "particlecloudwithweights",
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:1259]:
```suggestion
    "particle_cloud",
```

**Comment by shivaang12** [comment]
LGTM! Thanks for the PR @naiveHobo

**Comment by SteveMacenski** [comment]
Waiting on CI

**Comment by rfzeg** [comment]
> Publishing the particle poses and weights together would require a custom message. 
> 
> Another way would be to publish the weights and poses as separate messages on different topics. 

Leaving this comment for future reference:

Mayfield Robotics solved this is setting the Pose height value as the particle weight.
Benefit:  does not require a custom message or different topic.
Taken from here: http://docs.freekuri.com/reference/ros-nodes/laser-amcl.html

**Parameters**

**draw_weight_as_height*, booean (default: false)

If true, the height of the poses in the geometry_msgs::PoseArray broadcast on /particlecloud will have height (z-values) set to be proportional to the particle weight (highest particle at 1 m).

This could certainly not be used when localizing a robot moving in 3D.

**Comment by SteveMacenski** [comment]
> Mayfield Robotics solved this setting the Pose height value as the particle weight.

That's not a scalable hack. That might work for an internal company because they're not exposing it to thousands of users, but abusing messages is not a good solution.

**Comment by SteveMacenski** [comment]
I haven't seen those docs before, thanks for sharing. The lost detector page gave me good feels because that's exactly how I independently built the same tool at another company. Many of the other tools there are strong analogs in open source or exactly in open source which is nice to see and have a reference for

**Comment by mikeferguson** [comment]
> Mayfield Robotics solved this setting the Pose height value as the particle weight.

@SteveMacenski - suggest maybe adding this to the ticket advocating for the new message - as an example of what people are doing in the real world to work around the lack of a message

**Pull request merged**